src/core/ruby_version.rs

```diff
@@ -277,17 +285,28 @@ macro_rules! get_stack_trace(
             ruby_global_symbols_address_location: Option<usize>,
             source: &T,
             pid: Pid,
-        ) -> Result<StackTrace, MemoryCopyError> {
+            on_cpu: bool,
+        ) -> Result<Option<StackTrace>, MemoryCopyError> {
             let thread: $thread_type = get_execution_context(ruby_current_thread_address_location, ruby_vm_address_location, source)
                 .context(ruby_current_thread_address_location)?;
 
+            // testing the thread state in the interpreter.
+            if on_cpu && get_thread_status(&thread, source)? != 0 /* THREAD_RUNNABLE */ {
```

Comment:
I don't know ruby well enough - is this the equivalent of CPython's GIL? If so, it's probably better to add this as another option (compare to py-spy's `--active` for on-CPU, and `--gil` for GIL; 2 different options that can go together, but I might want `--active` only w/o `--gil`)

Reply:
My understanding is that they're very similar. Only one thread per VM can be executing ruby code at a time. Native C extensions can explicitly release the GVL, though, so that ruby and C code can execute in parallel when it's safe to do so. The situation is changing a bit in ruby 3 with the addition of ractors, which have their own lock and will probably need different logic for on/off CPU analysis. rbspy only has basic support for ractors right now (only looks at the main one), so we may need to defer that work until later.

I'm not opposed to introducing a `--gvl` option if we can find a reliable way to detect which thread holds the lock, but I'd recommend doing that in a future PR so that we can keep this one small-ish and focused. We might also want an `--off-cpu` option. Lots of future paths to explore here.

It seems that this `get_thread_status` check is the key mechanism, and the earlier check using remoteprocess's `Thread::active` is a fast/cheap filter to skip `get_trace` if we know that none of the process's threads is active (from the OS's perspective). Is that accurate?

Reply:
Thanks for the brief Ruby internals :)

> I'm not opposed to introducing a --gvl option if we can find a reliable way to detect which thread holds the lock,

I might have missed that - isn't it what we expect `THREAD_RUNNABLE` to tell? Or does runnable mean "can be run given that it obtains the GVL"? If the latter, then I think that looking at it doesn't make any sense at all.

> We might also want an --off-cpu option. Lots of future paths to explore here.

Yup. Well, basically, no `--on-cpu` means `--off-cpu` now :)

> It seems that this get_thread_status check is the key mechanism, and the earlier check using remoteprocess's Thread::active is a fast/cheap filter to skip get_trace if we know that none of the process's threads is active (from the OS's perspective). Is that accurate?

To my understanding - not precisely, I'll give an elaborated answer on the other discussion :sweat_smile:

Reply:
> I might have missed that - isn't it what we expect THREAD_RUNNABLE to tell? Or does runnable mean "can be run given that it obtains the GVL"? If the latter, then I think that looking at it doesn't make any sense at all.

I think this is a key question, and I don't know enough about RMI VM internals to give a confident answer. What you said seems reasonable. I'll poke around in some reference materials that I've collected and will post again if I find anything.

> Yup. Well, basically, no --on-cpu means --off-cpu now :)

My feeling is that the absence of `--on-cpu` should cause rbspy to collect all traces, which is the current shipping behavior (and what this PR would do in its current state, IIUC), and then we could introduce `--off-cpu` to do the opposite filtering that `--on-cpu` does, i.e. only collect traces if the thread is waiting.